[FSDP] Add `keep_low_precision_grads` support when CPU offloading

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #86513 [FSDP] Change `backward_prefetch` default to `BACKWARD_PRE`
* #86512 [FSDP] Add `_low_prec` prefix to param and reduce dtype varnames
* **#86495 [FSDP] Add `keep_low_precision_grads` support when CPU offloading**
* #85738 [FSDP] Add initial `summon_full_params(with_grads=True)`
* #84911 [FSDP] Add `use_orig_params`

When CPU offloading, FSDP uses `_cpu_grad`, not `_saved_grad_shard`. This adds support for `keep_low_precision_grads` for that case.

Commit:
[FSDP] Add `keep_low_precision_grads` support when CPU offloading

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `keep_low_precision_grads` support when CPU offloading"


When CPU offloading, FSDP uses `_cpu_grad`, not `_saved_grad_shard`. This adds support for `keep_low_precision_grads` for that case.

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -609,8 +613,9 @@ def test_mixed_precision_resnet(self):
-    def test_grads_reduced_precision(self):
-        self._test_grads_reduced_precision()
+    @parametrize("offload_params", [False, True])
+    def test_grads_reduced_precision(self, offload_params: bool):
```
with `self.subTest`?

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -944,6 +948,13 @@ def prepare_gradient_for_optim(self):
+                assert flat_param.grad is not None  # mypy
+                # This cast is meaningful when `param_dtype` is a low precision
+                # dtype.
+                flat_param.grad.data = flat_param.grad.to(self._config.param_dtype)
```
this is the mixed precision type, right? it is not fully clear based on the name, perhaps we should improve the naming?

awgu replied:
Let me rename this in a follow-up.

rohan-varma reviewed (approved):
LGTM

Commit:
Update on "[FSDP] Add `keep_low_precision_grads` support when CPU offloading"


When CPU offloading, FSDP uses `_cpu_grad`, not `_saved_grad_shard`. This adds support for `keep_low_precision_grads` for that case.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
